### PR TITLE
[FIXED JENKINS-26637] Allow regexp applied for matrix children logs

### PR DESCRIPTION
**Pull request by ikedam**
[JENKINS-26637](https://issues.jenkins-ci.org/browse/JENKINS-26637)

Based on #20 
Replaces #17 , #18.

Added an option for applying the regular expression to logs of matrix children.
screenshot

### CODE REVIEW

**Comment by ikedam** [comment]
* The default values for "Test regular expression for the matrix parent" is `true` for projects configured AFTER upgraded to 1.16.
    * As it's considered to be the common usage.
* The initial values for "Test regular expression for the matrix parent" is `false' for projects configured BEFORE upgraded to 1.16.
    * For backward compatibilities.

File: src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisherScheduleAction.java [hunk 1]
```diff
@@ -60,24 +69,11 @@ public boolean shouldSchedule(@Nonnull Run<?, ?> run, @Nonnull TaskListener list
...
-                    // If parseLog returns false, we didn't find the regular expression,
-                    // so return true.
-                    if (!parseLog(run.getLogFile(), regexpForRerun)) {
-                        LOGGER.log(Level.FINEST, "regexp not in logfile");
-                        return false;
-                    }
-                } catch (IOException e) {
-                    e.printStackTrace(listener
-                                      .error("error while parsing logs for naginator - forcing rebuild."));
-                }
```
Comment (ikedam) [inline src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisherScheduleAction.java]: Extracted to `testRegexp`. (as used also in `shouldScheduleForMatrixRun`)

File: src/main/resources/com/chikli/hudson/plugin/naginator/NaginatorPublisher/config.jelly [hunk 2]
```diff
@@ -17,13 +17,17 @@
...
-        </f:entry>
+            <f:entry title="${%Regular expression to search for}" field="regexpForRerun">
+                <f:textbox />
+            </f:entry>
+            <j:if test="${descriptor.matrixProject}">
+                <f:entry title="${%Test regular expression for the matrix parent}" field="regexpForMatrixParent">
+                    <f:checkbox />
+                </f:entry>
+            </j:if>
+        </f:optionalBlock>
```
Comment (ikedam) [inline src/main/resources/com/chikli/hudson/plugin/naginator/NaginatorPublisher/config.jelly]: Changes:
* "Regular expression to search for" is displayed only when you checked "Only rerun build if regular expression is found in output".
* Added "Test regular expression for matrix parent", which is displayed only for matrix projects.

File: src/test/java/com/chikli/hudson/plugin/naginator/NaginatorListenerTest.java [hunk 3]
```diff
@@ -45,46 +42,6 @@ protected void tearDown() throws Exception {
...
-        public static final class DescriptorImpl extends Descriptor<Builder> {
-            public String getDisplayName() {
-                return "MyBuilder";
-            }
-            public MyBuilder newInstance(StaplerRequest req, JSONObject data) {
-                return new MyBuilder("foo", Result.SUCCESS);
-            }
-        }
-    }
-
```
Comment (ikedam) [inline src/test/java/com/chikli/hudson/plugin/naginator/NaginatorListenerTest.java]: All changes in this file is to extract `MyBuilder` in order to make it usable from `NaginatorPublisherTest`

**Comment by ikedam** [comment]
Added following 5 test cases:
* Regular expression is applicable for matrix parents (2 cases: regular expression maches and doesn't matches).
* Regular expression is applicable for matrix children.
* The configuration page works (2 cases: for matrix-project and freestyle-project).

**Comment by thabok** [comment]
I closed #17 as it is now obsolete. Thanks for your efforts, I would be great to see this PR merged as soon as possible =)

**Comment by ikedam** [comment]
A month passed. How's this request going?
I want a new release with this change (or at least #20) to continue the development of build-timeout-plugin.
Who's the current maintainer of this plugin?

@ndeloof @KostyaSha  Do you still maintain this plugin?

**Comment by KostyaSha** [comment]
@ikedam  i got bless to fix job properties and make release from @abayer . Then decided to help with reviews of your PRs as i see you do good things.

Now i'm just watching how @ndeloof maintaining plugins.

**Comment by ikedam** [comment]
OK, I'll contact @abayer and @ndeloof ,CCing jenkinsci-dev group.

**Comment by ikedam** [comment]
I found a ticket requesting this feature: https://issues.jenkins-ci.org/browse/JENKINS-26637

**Comment by ikedam** [comment]
I'll make a new release 0.16 this weekend.

**Pull request merged**